### ISSUE TITLE
Performance differences in working with Heap vs DirectByteBuffer-backed ImmutableRB

### ISSUE THREAD

**Issue by maciej** [RoaringBitmap/RoaringBitmap#114]
I'm getting a ~2-4x performance drop when working with `HeapByteBuffer` backed `ImmutableRoaringBitmap`s compared to ones backed by a `DirectByteBuffer`.
It seems unlikely for me that this can be only attributed to different performance of the `ByteBuffer` implementations (but I can see myself being wrong here!).

@lemire are you aware of this performance characteristic?

Running:
* JMH1.13
* JDK 1.8.0_101, VM 25.101-b13
* MacBookProc 15’ Late 2013, Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz, 16GiB .
* OSX 10.11.6

```
Benchmark                            (bufferType)           (dataset)  Mode  Cnt    Score   Error  Units
BackingBufferBenchmark.intersectAll          heap  wikileaks-noquotes  avgt   20   13.648 ± 0.100  us/op
BackingBufferBenchmark.intersectAll        direct  wikileaks-noquotes  avgt   20    6.250 ± 0.047  us/op
BackingBufferBenchmark.intersectAll          heap     weather_sept_85  avgt   20  108.748 ± 1.671  us/op
BackingBufferBenchmark.intersectAll        direct     weather_sept_85  avgt   20   39.418 ± 0.373  us/op
BackingBufferBenchmark.intersectAll          heap       dimension_033  avgt   20  422.748 ± 5.789  us/op
BackingBufferBenchmark.intersectAll        direct       dimension_033  avgt   20  102.691 ± 0.859  us/op
BackingBufferBenchmark.intersectAll          heap       census-income  avgt   20   24.322 ± 0.277  us/op
BackingBufferBenchmark.intersectAll        direct       census-income  avgt   20    7.409 ± 0.072  us/op
BackingBufferBenchmark.intersectAll          heap        uscensus2000  avgt   20    2.980 ± 0.024  us/op
BackingBufferBenchmark.intersectAll        direct        uscensus2000  avgt   20    3.064 ± 0.028  us/op
```

The benchmark code can be found [here](https://github.com/maciej/roaring-buffer-perf/blob/master/src/main/scala/me/maciejb/buffper/BackingBufferBenchmarkRunner.scala).

One thing I've noticed looking at code of RoaringBitmaps is that `BufferUtil.isBackedBySimpleArray` is used in a couple of places to branch to two different code paths depending on the backing buffer type. Do both code paths have similar performance?

**Comment by lemire** [comment]
I am not sure it is a valid issue. 

In Java, arrays are much faster than Buffers. When Roaring can, it tries to bypass the Buffer and work with the underlying array for much better performance. When there is no underlying Java array... we are stuck with the Buffer interface and its terrible performance... but that's not our doing.

Now, of course, if you know how to optimize the code better...

**Comment by maciej** [comment]
@lemire I think there is some misunderstanding.

The `DirectByteBuffer`, to my knowledge is not backed by an array, however it's 2-4 *faster* than the array-backed `HeapByteBuffer`.

**Comment by lemire** [comment]
@maciej 

I am not sure that there is a misunderstanding. 


The code path following a successful  ``BufferUtil.isBackedBySimpleArray`` check will be massively faster than going through ``HeapByteBuffer``, but the test you provided won't use this code path. (This code path is meant for the ``MutableRoaringBitmap`` class.)

Though you are correct that ``HeapByteBuffer`` is backed by an array, that's not what I alluded to in my earlier comment and it is not what ``BufferUtil.isBackedBySimpleArray`` checks for. It is a ``byte[]`` array, but we work with either ``LongBuffer`` or ``ShortBuffer`` extracted from the ``ByteBuffer`` and those won't be array-backed. 

To be clear... this will print ``false``:

```
     LongBuffer x = ByteBuffer.allocate(8).asLongBuffer();
     System.out.println(x.hasArray());
```



So, if you provide a ``HeapByteBuffer``, you are stuck with code that looks like this...

```
 public ByteBuffer putLong(int i, long x) {
        Bits.putLong(this, ix(checkIndex(i, 8)), x, bigEndian);
        return this;
}
```

http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/nio/HeapByteBuffer.java#HeapByteBuffer.putLong%28long%29


This sort of code is clearly not meant for performance. There is nothing that ``RoaringBitmap`` can do about it... short of testing for the buffer type and providing an alternative implementation... and this alternative implementation would need to implement the same method Java offers, only with massively better performance... That might prove difficult.

 

Why does a ``DirectByteBuffer`` fare better? Because while it takes longer to do the allocation, once it is done, it uses better optimization *within the JVM*. This is beyond our control.


You can test this independently irrespective of ``RoaringBitmap``:

Grab

```bash
// grab https://github.com/lemire/microbenchmarks
$  mvn clean install
$ java -cp target/microbenchmarks-0.0.1-jar-with-dependencies.jar me.lemire.microbenchmarks.bytebuffer.DirectVSHeapVSArray
Benchmark                                  Mode  Samples        Score        Error  Units
m.l.m.b.DirectVSHeapVSArray.arrayTest     thrpt        5  2390647.053 ± 744114.493  ops/s
m.l.m.b.DirectVSHeapVSArray.directTest    thrpt        5  1234855.882 ± 693014.209  ops/s
m.l.m.b.DirectVSHeapVSArray.heapTest      thrpt        5   188097.091 ±  89827.964  ops/s
```


So, basically, this issue is "HeapByteBuffer" is slower than "DirectByteBuffer". Yes. But that's not our doing.

**Comment by lemire** [comment]
To put it another, I suggest you run other read/write tests on direct vs. heap ByteBuffer using Java types like ``long`` and measure the difference. I submit to you that they are not on par performance-wise. It is in this sense that I don't think this issue is valid. You should not expect the same performance from Roaring if you provide a backing buffer that has a different performance.

**Comment by lemire** [comment]
@maciej  Let me know if you agree with my answer. If so, I will close this issue.

**Comment by maciej** [comment]
@lemire absolutely! It's my bad that I haven't notice the fact that Roaring uses through the "buffer" branch whenever a `HeapByteBuffer` is passed. And clearly 4x perf drop makes sense then!
Thank you for your help on that one!

I'm closing it.
